Add a test for filling M2 from the database

arnout opened the pull request:
To be able to do this test, we need the autoconfig renew message to be handled in the agent, so add that as well.

On the way, some other small issues were found in the agent. These are fixed in separate commits.

arnout committed:
agent: slave: improve dummy radioBasicCaps generation

Set the operating class based on the radio band.

Set only one operating class. Before, one operating class was added for
each supported BSS, but that makes no sense.

Set supported bsss to 2 instead of 1, so we can test configuring
multiple SSIDs.

Clean up the TODO comments.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: slave: parse bss_type out of M2

In autoconfig_wsc_parse_m2_encrypted_settings(), parse the bss_type out
of the MultiAP IE in the WFA vendor extension of config data. Parse this
as individual booleans instead of a bitmask, since we'll need the
boolean values in the end.

For now, no check is done on the reserved bits. A warning is printed if
the backhaul STA bit is set.

Print the backhaul/fronthaul/teardown information in the debug log.
While we're at it, also put all this information on a single line. This
will make is easier for the tests to check if the info is correct by
grepping for it.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: slave: fix the parsing of ssid from M2

If ssid is empty, the ssid() function will give an error and return a
nullptr. So this situation has te be handled explicitly.

While we're at it, also check if it is smaller than the maximum allowed
length.

Both of these really should be done in tlvf itself.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: handle AP_AUTOCONFIGURATION_RENEW

In Multi-AP, the AP Autoconfiguration Renew message should trigger a
renew on all radios (even if the band only indicates one specific
radio). Therefore, this message is handled in the slave threads and not
in the backhaul manager.

However, the backhaul manager is the only one that can check the source
address, and this really needs to be done since controller discovery is
not repeated. Therefore, this check is done in backhaul manager, but
return false to make sure it is still forwarded to the slaves.

To handle the renew, just jump to the JOIN_MASTER state which will send
out the M1.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
tools: test_flows: add test of M2 contents

Add part of the ap_config test flow that checks that the M2 contents is
correct. Set some BSS info on the controller, trigger an autoconfig
renew, and check that the agent receives the expected SSID and
fronthaul/backhaul bits.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3702,44 +3702,45 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
             return false;
         }
 
+        // TODO Move APRadioBasicCapabilities creation to a separate function
```
Can we do that as part of this PR? It can even be a separate commit which moves all the big logic out of the case. At the first step I'll even be happy with a single function `start_autoconfiguration` and move all the logic there.

arnout replied:
> Can we do that as part of this PR? It can even be a separate commit which moves all the big logic out of the case.

I felt that this PR was already big enough as it is :-) Note that this comment was already there, I just moved it to the appropriate place.

>  At the first step I'll even be happy with a single function `start_autoconfiguration` and move all the logic there.

I'll stick with creating a function just for this tlv, similar to `autoconfig_wsc_add_m1()`

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -3702,44 +3702,45 @@ bool slave_thread::slave_fsm(bool &call_slave_select)
+        operationClassesInfo->operating_class()            = operating_class;
+        operationClassesInfo->maximum_transmit_power_dbm() = 0;
+
+        // TODO - the number of statically non operable channels can be 0 - meaning it is
```
Issue #8 is closed - so this TODO comment can be removed - TLVF has support for variable-length lists with zero-length.
IMO the only reason to keep a non-operable channel is to test the controller code which sets the supported channels based on the operating class, but it has been tested so I wouldn't bother.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4510,19 +4520,23 @@ bool slave_thread::handle_autoconfiguration_wsc(Socket *sd, ieee1905_1::CmduMess
         sMacAddr bssid;
         WSC::eWscAuth authtype;
         WSC::eWscEncr enctype;
+        bool fronthaul;
```
(nitpick) I prefer to keep multiple variables definitions of the same type in the same line:
```
bool fronthaul, backhaul, teardown;
```

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4415,6 +4415,20 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
+        LOG(INFO) << "SSID too long: " << config_data.ssid_length();
+        return false;
+    }
+    // TODO tlvf should do conversion to std::string
```
See #265 - you can take over if needed or we can merge it and you can rebase.

arnout replied:
I'll see after #265 has been merged.

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4415,6 +4415,20 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
     // Swap back to host byte order to read and use config_data
     config_data.class_swap();
+
+    // TODO tlvf should check this
```
Correct - we should add a `_length_max` meta variable to TLVF and add this support. Can you open a task for it and refer to it in the comment?

tomereli replied:
NVM - I'm adding support for this as well in #265

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4443,6 +4448,53 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
     return true;
 }
 
+/**
```
Copy paste bug

tomereli commented on agent/src/beerocks/slave/son_slave_thread.cpp:
```diff
@@ -4443,6 +4448,53 @@ bool slave_thread::autoconfig_wsc_parse_m2_encrypted_settings(
+    auto tlvAlMac = cmdu_rx.addClass<ieee1905_1::tlvAlMacAddressType>();
+    if (tlvAlMac) {
+        LOG(DEBUG) << "tlvAlMac=" << network_utils::mac_to_string(tlvAlMac->mac());
+        // TODO register/update mapping of AL-MAC to interface
```
Do we have a task for this? if so, it would be useful to add the link here.

arnout committed:
agent: slave: move add_radio_basic_capabilities to a separate function

Constructing the AP Radio Basic Capabilities TLV is quite complicated,
so it makes sense to put it in a separate function. In addition, this
TLV is also needed for the AP Capability Report message, so there will
be a second caller once that is implemented.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: slave: move comment to the right place

Also rewrap it to 100 columns.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>

arnout committed:
agent: slave: add_radio_basic_capabilities(): remove dummy non-operable channel

When generating the (dummy) radio basic capabilities TLV, a statically
non-operable channel was added due to isse #8. However, that issue was
fixed long ago, so it is no longer necessary to do this.

Signed-off-by: Arnout Vandecappelle (Essensium/Mind) <arnout@mind.be>
